File: py/y2024/day1/day1.py
import sys
import re
import collections

def parse(input: str):
    lines = input.split("\n")
    lines2 = [re.split(r'\s+', line) for line in lines]
    return lines2


def part1(input: str):
    lines = parse(input)
    left = sorted(int(line[0]) for line in lines)
    right = sorted(int(line[1]) for line in lines)

    sum = 0
    for l,r in zip(left, right):
        diff = abs(l - r)
        sum += diff

    return sum

def part2(input: str):
    lines = parse(input)
    left = (int(line[0]) for line in lines)
    right = (int(line[1]) for line in lines)

    occurences = collections.Counter(right)
    print(occurences)

    sum = 0
    for l in left:
        occ = occurences.get(l, 0)
        val = l * occurences.get(l, 0)
        sum += val

    return sum

if __name__ == "__main__":
    input = sys.stdin.read()
    result = part2(input)
    print("diff", result)

File: vite/src/y2024/day1/day1.ts
import { splitIntoLines } from "../../util";

export function part1(input: string) {
    const lines = splitIntoLines(input);

    let sum = 0;
    const left: number[] = [];
    const right: number[] = [];

    for (const line of lines) {
        const [a, b] = line.split(/\s+/).map(Number);
        left.push(a);
        right.push(b);
    }
    
    left.sort((a, b) => a - b);
    right.sort((a, b) => a - b);

    for (let i = 0; i < left.length; i++) {
        const a = left[i];
        const b = right[i];

        const diff = Math.abs(a - b);

        sum += diff;
    }

    return sum;
}

export function part2(input: string) {
    const lines = splitIntoLines(input);

    const left: number[] = [];
    const right: number[] = [];
    let sum = 0;

    for (const line of lines) {
        const [a, b] = line.split(/\s+/).map(Number);
        left.push(a);
        right.push(b);
    }

    const occurences = new Map<number, number>();
    for (const num of right) {
        occurences.set(num, (occurences.get(num) ?? 0) + 1);
    }

    for (const num of left) {
        const count = occurences.get(num) ?? 0;
        sum += count * num;
    }

    return sum;
}
